src/history_be/models/stable_bucket_list.mo

```diff
@@ -0,0 +1,179 @@
+    if (bucketIndex >= l.bucketsAllocated) {
+      Prim.trap("Cannot retrieve values of bucket #" # debug_show bucketIndex # ". Bucket was not allocated");
+    };
+    var ptr = _loadFirstRecordPtr(l, bucketIndex);
+    {
+      next = func() : ?Blob {
+        if (ptr == 0) return null;
+        let { nextPtr; data } = _loadRecord(l, ptr);
+        ptr := nextPtr;
+        ?data;
+      };
+    };
+  };
+
+  public func valuesRev(l : StableBucketList, bucketIndex : Nat64) : Iter.Iter<Blob> {
+    if (bucketIndex >= l.bucketsAllocated) {
+      Prim.trap("Cannot retrieve valuesRev of bucket #" # debug_show bucketIndex # ". Bucket was not allocated");
+    };
+    var ptr = _loadLastRecordPtr(l, bucketIndex);
+    {
+      next = func() : ?Blob {
+        if (ptr == 0) return null;
+        let { prevPtr; data } = _loadRecord(l, ptr);
+        ptr := prevPtr;
+        ?data;
+      };
+    };
+  };
+
+  public func memoryStats(l : StableBucketList) : {
+    pages : { indexTable : Nat64; data : Nat64 };
+    bytesUsed : Nat64;
+  } = {
+    pages = {
+      indexTable = Region.size(l.indexTable);
+      data = Region.size(l.data);
+    };
+    bytesUsed = l.dataLength;
+  };
+
+  // ======================== INTERNAL PRIVATE FUNCTIONALITY ========================
+  private let MAX_BUCKET_INDEX : Nat64 = 16_777_215; // 2^24 - 1
+
+  private func _loadListLength(l : StableBucketList, bucketIndex : Nat64) : Nat64 = Region.loadNat64(l.indexTable, bucketIndex * 3 * 8);
+  private func _storeListLength(l : StableBucketList, bucketIndex : Nat64, v : Nat64) = Region.storeNat64(l.indexTable, bucketIndex * 3 * 8, v);
+
+  private func _loadFirstRecordPtr(l : StableBucketList, bucketIndex : Nat64) : Nat64 = Region.loadNat64(l.indexTable, (bucketIndex * 3 + 1) * 8);
+  private func _storeFirstRecordPtr(l : StableBucketList, bucketIndex : Nat64, v : Nat64) = Region.storeNat64(l.indexTable, (bucketIndex * 3 + 1) * 8, v);
+
+  private func _loadLastRecordPtr(l : StableBucketList, bucketIndex : Nat64) : Nat64 = Region.loadNat64(l.indexTable, (bucketIndex * 3 + 2) * 8);
+  private func _storeLastRecordPtr(l : StableBucketList, bucketIndex : Nat64, v : Nat64) = Region.storeNat64(l.indexTable, (bucketIndex * 3 + 2) * 8, v);
+
+  private func _loadRecord(l : StableBucketList, pointer : Nat64) : Record {
+    let size = Region.loadNat16(l.data, pointer);
+    let prevPtr = Region.loadNat64(l.data, pointer + 2);
+    let nextPtr = Region.loadNat64(l.data, pointer + 10);
+    let data = Region.loadBlob(l.data, pointer + 18, Nat16.toNat(size));
+    { data; prevPtr; nextPtr };
+  };
+
+  private func _appendRecord(l : StableBucketList, record : Record) : Nat64 {
+    let recordSize : Nat64 = Nat64.fromNat(record.data.size()) + 18;
+    let pointer = l.dataLength;
+    if (65536 * Region.size(l.data) - pointer < recordSize and Region.grow(l.data, 2) == 0xFFFF_FFFF_FFFF_FFFF) {
+      Prim.trap("Out of memory");
+    };
```

Comment:
Since we are using this in multiple places, we can put this in a function.

Reply:
We use it now in two places only: once for index table and once for data region. So I think it's more readable if we don't create a dedicated function for that